balancer.go

```diff
@@ -0,0 +1,329 @@
+	forwarderMap map[string]*Forwarder
+	mutex        sync.Mutex
+}
+
+// NewLoadBalancer creates new instance of the load balancer
+// using array of pool configuration. For each pool it is
+// assumed that it has unique port, otherwise Listen will
+// fail with the error
+// TODO: Add checkup in constructor for enforce unique port per pool configuration
+func NewLoadBalancer(ctx context.Context, cfgPool []ServicePool, opt Options) (*LoadBalancer, error) {
+	id, err := uuid.NewRandom()
+	if err != nil {
+		return nil, fmt.Errorf("gen id failed")
+	}
+	if ctx == nil {
+		return nil, fmt.Errorf("missing parameter ctx")
+	}
+	if cfgPool == nil || len(cfgPool) == 0 {
+		return nil, fmt.Errorf("missing parameter cfgPool")
+	}
+
+	logger := newZeroLogForName("xlb", id.String(), opt.LogLevel)
+	if opt.Logger != nil {
+		logger = *opt.Logger
+	}
+
+	poolMap := make(map[string]ServicePool)
+	for _, pool := range cfgPool {
+		if len(pool.Identity()) == 0 {
+			return nil, fmt.Errorf("pool missing identity")
+		}
+		poolMap[pool.Identity()] = pool
+	}
+
+	derCtx, cancelFunc := context.WithCancel(ctx)
+	return &LoadBalancer{
+		id:           id.String(),
+		runCtx:       derCtx,
+		killCtx:      cancelFunc,
+		logger:       logger,
+		forwarderMap: map[string]*Forwarder{},
+		poolMap:      poolMap,
+	}, nil
+}
+
+// UpdatePool will update pool using pool.Identity() method, this will
+// trigger hot-swap operation on running forwarder for pool and should
+// replace targets behind the load balancer without the restart
+func (lb *LoadBalancer) UpdatePool(pool ServicePool) error {
+	if len(pool.Identity()) == 0 {
+		return fmt.Errorf("pool missing identity")
+	}
+	lb.mutex.Lock()
+	defer lb.mutex.Unlock()
+	lb.poolMap[pool.Identity()] = pool
+	if fwd, exists := lb.forwarderMap[pool.Identity()]; exists {
+		fwd.UpdateServicePool(pool)
+	}
+	return nil
+}
+
+// Listen will try to launch balancer on all the required ports, strategy is all or nothing
+func (lb *LoadBalancer) Listen() error {
+
+	// Do this step to ensure that we will fail on misconfiguration if more than
+	// one service pool mapping presented to this load balance for mTLS
+	// TODO: TLS Allows manual verification for the handshake by that we can launch multiple pools on the same port
+	// TODO: for the first version we will limit 1 pool per port for simplicity
+	mapping, err := collectListenTargets(lb.poolMap)
+	if err != nil {
+		return err
+	}
+
+	type schedule struct {
+		port int
+		tls  *tls.Config
+		pool ServicePool
+	}
+	scheduleListeners := make([]schedule, len(mapping))
+	i := 0
+	// Build schedule list not to have thread failures at this stage
+	for port, identity := range mapping {
+
+		pki, err := tlsutil.FromPKI(identity.TLSData().GetCertificate(), identity.TLSData().GetPrivateKey())
+		if err != nil {
+			return fmt.Errorf("invalid service pool pki data")
+		}
+
+		config := &tls.Config{
+			Certificates: []tls.Certificate{pki.Certificate},
+			ClientAuth:   tls.RequestClientCert,
```

Comment:
Does `RequestClientCert` client AUTH  is sufficient for mTLS ?

Reply:
No, thanks for catching the typo there. It was changed for certificate gen+matching phase, and was not changed back.